File: PythonLeetcode/LeetCodeE/513_FindBottomLeftTreeValue.py
# 513. Find Bottom Left Tree Value


# Given a binary tree, find the leftmost value in the last row of the tree.

# Example 1:
# Input:

#    2
#   / \
#  1   3

# Output:
# 1

# Example 2: 
# Input:

#        1
#       / \
#      2   3
#     /   / \
#    4   5   6
#       /
#      7

# Output:
# 7
# Note: You may assume the tree (i.e., the given root node) is not NULL.

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class BottomLeftTreeValue(object):

    def doit1(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.mostleft = (root.val, 0)
        def search(node, depth):
            
            if not node.left and not node.right and depth > self.mostleft[1]:
                self.mostleft = (node.val, depth)

            if node.left:
                search(node.left, depth + 1)
            if node.right:
                search(node.right, depth + 1)

        search(root, 0)
        return self.mostleft[0]
            
    # bfs
    def doit(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        res = (root.val, 0)
        bfs = [(root, 0)]

        while bfs:

            N = len(bfs)
            for _ in range(N):
                c, d = bfs.pop(0)
                if not c.left and not c.right and d > res[1]:
                    res = (c.val, d)
                
                if c.left:
                    bfs.append((c.left, d + 1))

                if c.right:
                    bfs.append((c.right, d + 1))

        return res[0]


    def doit(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        bl = (0,0,root)
        bfs = [bl]
        
        while bfs:
            depth, left_num, node = bfs.pop()

            if node.left or node.right:

                if node.left:
                    bfs.append((depth+1, left_num+1, node.left))

                if node.right:
                    bfs.append((depth+1, left_num, node.right))
            else:
                #do the comparison, if any one of depth and letf_num is higher, replace
                bl_depth, bl_left, _  = bl

                if (depth > bl_depth) or  ((depth == bl_depth) and (left_num > bl_left)):
                    #replace
                    bl = (depth, left_num, node)

        return bl[2].val      

if __name__ == "__main__":

    root = TreeNode(1)
    root.right = TreeNode(3)
    root.left = TreeNode(2)
    root.left.left = TreeNode(4)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(6)
    root.right.left.left = TreeNode(7)

    res = BottomLeftTreeValue().doit(root)

    

File: cppLeetcode/leetcodeM/513_FindBottomLeftTreeValue.cpp
/*
513. Find Bottom Left Tree Value

Given the root of a binary tree, return the leftmost value in the last row of the tree.

Example 1:


Input: root = [2,1,3]
Output: 1
Example 2:


Input: root = [1,2,3,4,null,5,6,null,null,7]
Output: 7
 

Constraints:

The number of nodes in the tree is in the range [1, 104].
-231 <= Node.val <= 231 - 1

*/

#include <functional>
#include <stack>

using std::stack;


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BottomLeftValue {


    struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    };

public:

    int doit_dfs(TreeNode* root) {
        
        int maxdepth = -1, val = INT_MAX;
        
        std::function<void(TreeNode*, int)> dfs = [&] (TreeNode* p, int depth) {
            
            if (!p) return;
            
            if (depth > maxdepth) {
                maxdepth = depth;
                val = p->val;
            }
            
            
            dfs(p->left, depth+1);
            dfs(p->right, depth+1);
        };
        
        dfs(root, 0);
        return val;
    }

    int doit_dfs(TreeNode* root) {

        stack<std::pair<TreeNode*, int>> st;
        st.push({root, 0});

        int maxdepth = -1, val = INT_MAX;

        while (!st.empty()) {

            auto it = st.top();
            TreeNode* cur = it.first;
            int depth = it.second;
            st.pop();

            if (depth > maxdepth) {
                maxdepth = depth;
                val = cur->val;
            }
            
            if (cur->right) {
                st.push({cur->right, depth+1});
            }

            if (cur->left) {
                st.push({cur->left, depth+1});
            }
        }

        return val;
    }

};